File: app/src/main/java/com/example/photoorganizer/data/repository/FolderRepository.kt
package com.example.photoorganizer.data.repository

import android.net.Uri
import com.example.photoorganizer.data.local.database.dao.FolderDao
import com.example.photoorganizer.data.local.database.entities.FolderEntity
import com.example.photoorganizer.data.local.database.entities.LearningStatus
import com.example.photoorganizer.data.local.saf.SafDataSource
import com.example.photoorganizer.domain.model.Folder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for folder operations following offline-first pattern.
 *
 * The repository acts as a single source of truth for folder data:
 * - Local Room database is the source of truth
 * - SAF discovery provides the data source
 * - Changes are synced from SAF → Local DB → UI
 *
 * @property folderDao DAO for folder database operations
 * @property safDataSource Data source for SAF folder discovery
 */
@Singleton
class FolderRepository @Inject constructor(
    private val folderDao: FolderDao,
    private val safDataSource: SafDataSource
) {

    /**
     * All folders as a reactive Flow.
     * Emits whenever folders change in the database.
     */
    val folders: Flow<List<Folder>> = folderDao.getAll()
        .map { entities -> entities.map { it.toDomainModel() } }

    /**
     * Active folders only (included in organization).
     */
    val activeFolders: Flow<List<Folder>> = folderDao.getActive()
        .map { entities -> entities.map { it.toDomainModel() } }

    /**
     * Discovers folders via SAF and syncs them to local database.
     * Implements offline-first pattern: SAF is data source, local DB is truth.
     *
     * @param picturesUri The Pictures/ directory URI to discover folders within
     * @return Result.success(Unit) on success, Result.failure(exception) on error
     */
    suspend fun discoverAndSyncFolders(picturesUri: Uri): Result<Unit> =
        withContext(Dispatchers.IO) {
            try {
                // 1. Discover folders from SAF
                val discoveredFolders = safDataSource.discoverFolders(picturesUri)

                // 2. Map to entities, preserving existing learning data
                val existingFolders = folderDao.getAll()
                    .map { list -> list.associateBy { it.uri } }
                    .firstOrNull() ?: emptyMap()

                val folderEntities = discoveredFolders.map { folderInfo ->
                    val existing = existingFolders[folderInfo.uri]
                    FolderEntity(
                        uri = folderInfo.uri,
                        name = folderInfo.name,
                        displayName = folderInfo.displayName,
                        photoCount = folderInfo.photoCount,
                        isActive = existing?.isActive ?: true,
                        learnedLabels = existing?.learnedLabels,
                        learningStatus = existing?.learningStatus ?: LearningStatus.PENDING.name,
                        createdAt = existing?.createdAt ?: System.currentTimeMillis()
                    )
                }

                // 3. Atomic sync: delete all, insert new
                folderDao.syncFolders(folderEntities)

                Result.success(Unit)
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    /**
     * Persists SAF URI permission for accessing the folder across reboots.
     *
     * @param uri The folder URI to persist permission for
     * @return Result.success(Unit) on success, Result.failure(exception) on error
     */
    suspend fun persistPermission(uri: Uri): Result<Unit> =
        withContext(Dispatchers.IO) {
            try {
                safDataSource.persistFolderPermission(uri)
                Result.success(Unit)
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    /**
     * Checks if persistable permission exists for the given URI.
     *
     * @param uri The URI to check
     * @return true if permission is persisted, false otherwise
     */
    suspend fun hasPermission(uri: Uri): Boolean =
        withContext(Dispatchers.IO) {
            safDataSource.hasPersistablePermission(uri)
        }

    /**
     * Updates the learning status for a folder.
     *
     * @param uri The folder URI
     * @param status The new learning status
     */
    suspend fun updateLearningStatus(uri: String, status: LearningStatus) {
        withContext(Dispatchers.IO) {
            folderDao.updateLearningStatus(uri, status.name)
        }
    }

    /**
     * Updates the learned labels for a folder.
     * Called after ML learning phase completes.
     *
     * @param uri The folder URI
     * @param labels Map of label → confidence score
     */
    suspend fun updateLearnedLabels(uri: String, labels: Map<String, Float>) {
        withContext(Dispatchers.IO) {
            val jsonLabels = labels.entries.joinToString(", ") { "${it.key}: ${it.value}" }
            // Note: In production, use proper JSON serialization (e.g., Gson or kotlinx.serialization)
            val jsonString = "{$jsonLabels}"
            folderDao.updateLearnedLabels(uri, jsonString)
        }
    }

    /**
     * Gets a folder by its URI.
     *
     * @param uri The folder URI
     * @return The folder if found, null otherwise
     */
    suspend fun getFolderByUri(uri: String): Folder? =
        withContext(Dispatchers.IO) {
            folderDao.getByUri(uri)?.toDomainModel()
        }

    /**
     * Toggles the active status of a folder.
     * Active folders are included in photo organization.
     *
     * @param uri The folder URI
     * @param isActive The new active status
     */
    suspend fun setFolderActive(uri: String, isActive: Boolean) {
        withContext(Dispatchers.IO) {
            folderDao.getByUri(uri)?.let { entity ->
                folderDao.update(entity.copy(isActive = isActive))
            }
        }
    }

    /**
     * Converts a FolderEntity to domain model.
     */
    private fun FolderEntity.toDomainModel(): Folder {
        return Folder(
            uri = uri,
            name = name,
            displayName = displayName,
            photoCount = photoCount,
            isActive = isActive,
            learningStatus = LearningStatus.valueOf(learningStatus),
            learnedLabels = parseLearnedLabels(learnedLabels),
            createdAt = createdAt
        )
    }

    /**
     * Parses JSON string of learned labels into a Map.
     * Note: In production, use proper JSON deserialization.
     */
    private fun parseLearnedLabels(json: String?): Map<String, Float> {
        if (json.isNullOrBlank()) return emptyMap()

        // Simple parsing: "{label1: 0.9, label2: 0.8}"
        return try {
            json.removeSurrounding("{", "}")
                .split(", ")
                .mapNotNull { pair ->
                    val parts = pair.split(": ")
                    if (parts.size == 2) {
                        parts[0] to parts[1].toFloatOrNull()
                    } else null
                }
                .filter { it.second != null }
                .associate { it.first to it.second!! }
        } catch (e: Exception) {
            emptyMap()
        }
    }
}

File: lib/data/repositories/folder_repository.dart
import 'package:flutter/foundation.dart';
import '../database/database_service.dart';
import '../platform/saf_service.dart';
import '../platform/result.dart';
import '../../domain/models/folder_model.dart';
import '../../domain/models/photo_model.dart';

/// Offline-first folder repository
/// Database is source of truth, SAF is data source
class FolderRepository {
  final DatabaseService databaseService;
  final SafService safService;

  FolderRepository({
    required this.databaseService,
    required this.safService,
  });

  /// Get folders as a reactive stream from database (source of truth)
  Future<List<FolderModel>> getFolders() async {
    return await databaseService.getFolders();
  }

  /// Discover folders via SAF and sync to database
  /// This is the offline-first sync operation:
  /// 1. Call SAF discovery
  /// 2. On success: delete existing folders, insert new ones
  /// 3. Update UI via database stream
  Future<Result<void>> discoverAndSyncFolders(String baseUri) async {
    try {
      // Step 1: Discover folders from SAF
      final discoverResult = await safService.discoverFolders(baseUri);
      
      if (discoverResult.isFailure) {
        return Result.failure(discoverResult.error ?? 'Unknown error');
      }

      final folders = discoverResult.data ?? [];
      debugPrint('Discovered ${folders.length} folders via SAF');

      // Step 2: Sync to database (delete old, insert new)
      final db = await databaseService.database;
      
      // Begin transaction for atomicity
      await db.transaction((txn) async {
        // Delete existing folders for this base URI
        // For now, we clear all and re-insert (simple approach)
        // In production, you might want to track which folders are from which base
        await txn.delete('folders');
        
        // Insert newly discovered folders
        for (final folder in folders) {
          await txn.insert(
            'folders',
            folder.toMap(),
            conflictAlgorithm: ConflictAlgorithm.replace,
          );
        }
      });

      debugPrint('Synced ${folders.length} folders to database');
      return Result.success(null);
    } catch (e) {
      debugPrint('Error syncing folders: $e');
      return Result.failure('Failed to sync folders: $e');
    }
  }

  /// Persist permission for a folder URI
  Future<Result<void>> persistFolderPermission(String uri) async {
    try {
      // First check if we already have permission
      final hasPermResult = await safService.hasPermission(uri);
      if (hasPermResult.isSuccess && (hasPermResult.data ?? false)) {
        debugPrint('Already have permission for $uri');
        return Result.success(null);
      }

      // Persist the permission
      final result = await safService.persistPermission(uri);
      if (result.isFailure) {
        return Result.failure(result.error ?? 'Failed to persist permission');
      }

      debugPrint('Persisted permission for $uri');
      return Result.success(null);
    } catch (e) {
      debugPrint('Error persisting permission: $e');
      return Result.failure('Failed to persist permission: $e');
    }
  }

  /// Check if we have permission for a folder URI
  Future<bool> hasPermission(String uri) async {
    try {
      final result = await safService.hasPermission(uri);
      return result.isSuccess && (result.data ?? false);
    } catch (e) {
      debugPrint('Error checking permission: $e');
      return false;
    }
  }

  /// Update learning status for a folder
  Future<void> updateLearningStatus(String uri, String status) async {
    await databaseService.updateFolderLearningStatus(uri, status);
    debugPrint('Updated learning status for $uri to $status');
  }

  /// Update learned labels for a folder
  Future<void> updateLearnedLabels(String uri, Map<String, double> labels) async {
    final db = await databaseService.database;
    await db.update(
      'folders',
      {
        'learned_labels': labels.toString(),
      },
      where: 'uri = ?',
      whereArgs: [uri],
    );
    debugPrint('Updated learned labels for $uri');
  }

  /// Get a specific folder by URI
  Future<FolderModel?> getFolderByUri(String uri) async {
    final folders = await getFolders();
    return folders.firstWhere((f) => f.uri == uri, orElse: () => null as FolderModel);
  }

  /// Count photos in a folder (uses cached count from database)
  Future<int> getPhotoCount(String uri) async {
    final folder = await getFolderByUri(uri);
    return folder?.photoCount ?? 0;
  }
}
